### ISSUE TITLE
Type "TResult" cannot be assigned to type "TSource"

### ISSUE THREAD

**Issue by dbrattli** [microsoft/pyright#1265]
I'm getting the error "TResult" cannot be assigned to type "TSource", which seems legit, but I still think there is a bug here

**Describe the bug**

The bug involves the use of an `identity` function `T -> T` where pyright  does not seem to match up the types correctly when used together with other functions.

**To Reproduce**

```python
from typing import Callable, Generic, List, TypeVar

TSource = TypeVar("TSource")
TResult = TypeVar("TResult")
TError = TypeVar("TError")
T = TypeVar("T")


def identity(expr: T) -> T:
    return expr


class Result(Generic[TSource, TError]):
    pass


class Ok(Result[TSource, TError]):
    def __init__(self, value: TSource):
        pass


def traverse(fn: Callable[[TSource], Result[TResult, TError]], lst: List[TSource]) -> Result[List[TResult], TError]:
    return Ok([])


def sequence(source: List[Result[TSource, TError]]) -> Result[List[TSource], TError]:
    """Execute a sequence of result returning commands and collect the
    sequence of their response."""

    return traverse(identity, source)
```

This gives type errors:

<img width="1103" alt="Screenshot 2020-12-11 at 08 21 47" src="https://user-images.githubusercontent.com/849479/101874875-f8766880-3b89-11eb-8341-4a9bbc7f32df.png">
<img width="1158" alt="Screenshot 2020-12-11 at 08 21 54" src="https://user-images.githubusercontent.com/849479/101874880-f9a79580-3b89-11eb-875b-43fbfc9ce729.png">
<img width="1170" alt="Screenshot 2020-12-11 at 08 22 00" src="https://user-images.githubusercontent.com/849479/101874889-fb715900-3b89-11eb-9960-40be463d88f0.png">

**Expected behavior**

I expect this code to type check correctly. Implementing the exact same code is F# gives not type errors:

```f#
// Learn more about F# at http://fsharp.org

type Result<'TSource, 'TResult>() =
    do ()

type Ok<'TSource, 'TResult>(value: 'TSource) =
    inherit Result<'TSource, 'TResult> ()

let identity(expr: 'T) : 'T =
    expr

let traverse(fn: 'TSource -> Result<'TResult, 'TError>, lst: list<'TSource>) : Result<list<'TResult>, 'TError> =
    Ok [] :> _

let sequence(source: list<Result<'TSource, 'TError>>) : Result<list<'TSource>, 'TError> =
    traverse(identity, source)
```

**VS Code extension or command-line**

vscode extension: 1.1.94

**Comment by erictraut** [comment]
I think pyright is doing the right thing here.

You are trying to pass the function `identity` as the first argument to the call to `traverse`. The first parameter of `traverse` is annotated to indicate that it must be a function that takes as a first parameter some generic type and it must return a Result object. The `identity` function doesn't match that signature. It does not return a Result object.

There are other problems in the code as well, but I'll stop there.

**Comment by dbrattli** [comment]
I still think it's a bug. The `identity` function matches. There is nothing that says that `TSource` cannot be `Result[TResult, TError]` in `traverse` and then the callable matches with an identity function `Callable[[Result[TResult, TError]], Result[TResult, TError]]`.

**Comment by dbrattli** [comment]
Also in Haskell: 
```hs
traverse :: Applicative f => (a -> f b) -> t a -> f (t b)

sequenceA :: Applicative f => t (f a) -> f (t a)
sequenceA = traverse id```
```
http://gensoft.pasteur.fr/docs/ghc/6.8.3/libraries/base/src/Data-Traversable.html

**Comment by erictraut** [comment]
It may work differently in other languages, but that's not how it works in Python's type system. When a Python type checker encounters a call to a generic function, it attempts to "solve" the type variables in the callee's signature. In the case of the `traverse` function, it attempts to find types that solve for `TSource`, `TResult` and `TError`. It doesn't attempt to solve for `T` because `T` doesn't appear within the signature of `traverse`. Because `T` contains no bound type, it is assumed that it can be any `object`, and that cannot be assigned to a `Result` object.

**Comment by dbrattli** [comment]
Are you saying that I must add specify overloads to the callee for every way I want to call it  for this to type check e.g:

```python
@overload
def traverse(
    fn: Callable[[Result[TResult, TError]], Result[TResult, TError]], lst: List[Result[TResult, TError]]
) -> Result[List[TResult], TError]:
    ...


def traverse(fn: Callable[[TSource], Result[TResult, TError]], lst: List[TSource]) -> Result[List[TResult], TError]:
    return Ok([])


def sequence(source: List[Result[TSource, TError]]) -> Result[List[TSource], TError]:
    """Execute a sequence of result returning commands and collect the
    sequence of their response."""

    return traverse(identity, source)

```

That works 😄

**Comment by erictraut** [comment]
This is how the Python type system works.

**Comment by dbrattli** [comment]
Thanks for clarifying. This knowledge will be very helpful for me going forwards.
